<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2 Introduction to the course | Introduction to evolutionary and coevolutionary theory</title>
  <meta name="description" content="This is the webpage for the course “Introduction to evolutionary and coevolutionary theory”. Here you will find the course schedule, as well as all of the lecture notes." />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="2 Introduction to the course | Introduction to evolutionary and coevolutionary theory" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the webpage for the course “Introduction to evolutionary and coevolutionary theory”. Here you will find the course schedule, as well as all of the lecture notes." />
  <meta name="github-repo" content="lgcosmo/evo_coevo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 Introduction to the course | Introduction to evolutionary and coevolutionary theory" />
  
  <meta name="twitter:description" content="This is the webpage for the course “Introduction to evolutionary and coevolutionary theory”. Here you will find the course schedule, as well as all of the lecture notes." />
  

<meta name="author" content="Leandro G. Cosmo" />


<meta name="date" content="2024-07-15" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="exercise-evolution-of-allele-frequencies.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Course notes</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.1</b> Registration</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i><b>1.2</b> Course schedule</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#bibliography"><i class="fa fa-check"></i><b>1.3</b> Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-the-course.html"><a href="introduction-to-the-course.html"><i class="fa fa-check"></i><b>2</b> Introduction to the course</a>
<ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-the-course.html"><a href="introduction-to-the-course.html#the-fundamental-concepts-of-evolution-by-natural-selection"><i class="fa fa-check"></i><b>2.1</b> The fundamental concepts of evolution by natural selection</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-the-course.html"><a href="introduction-to-the-course.html#natural-selection-and-the-evolution-of-allele-frequencies-within-populations"><i class="fa fa-check"></i><b>2.2</b> Natural selection and the evolution of allele frequencies within populations</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="introduction-to-the-course.html"><a href="introduction-to-the-course.html#building-our-first-equation"><i class="fa fa-check"></i><b>2.2.1</b> Building our first equation</a></li>
<li class="chapter" data-level="2.2.2" data-path="introduction-to-the-course.html"><a href="introduction-to-the-course.html#wrights-equation-for-the-adaptive-landscape-of-alleles"><i class="fa fa-check"></i><b>2.2.2</b> Wright’s equation for the adaptive landscape of alleles</a></li>
<li class="chapter" data-level="2.2.3" data-path="introduction-to-the-course.html"><a href="introduction-to-the-course.html#references"><i class="fa fa-check"></i><b>2.2.3</b> References</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="exercise-evolution-of-allele-frequencies.html"><a href="exercise-evolution-of-allele-frequencies.html"><i class="fa fa-check"></i><b>3</b> Exercise: Evolution of allele frequencies</a>
<ul>
<li class="chapter" data-level="3.1" data-path="exercise-evolution-of-allele-frequencies.html"><a href="exercise-evolution-of-allele-frequencies.html#simulating-evolutionary-changes-in-allele-frequencies"><i class="fa fa-check"></i><b>3.1</b> Simulating evolutionary changes in allele frequencies</a></li>
<li class="chapter" data-level="3.2" data-path="exercise-evolution-of-allele-frequencies.html"><a href="exercise-evolution-of-allele-frequencies.html#using-different-assumptions-for-the-fitness-of-alleles-a-and-b"><i class="fa fa-check"></i><b>3.2</b> Using different assumptions for the fitness of alleles A and B</a></li>
<li class="chapter" data-level="3.3" data-path="exercise-evolution-of-allele-frequencies.html"><a href="exercise-evolution-of-allele-frequencies.html#simulating-wrights-adaptive-landscape"><i class="fa fa-check"></i><b>3.3</b> Simulating Wright’s adaptive landscape</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="exercise-simulating-evolution-using-the-price-theorem.html"><a href="exercise-simulating-evolution-using-the-price-theorem.html"><i class="fa fa-check"></i><b>4</b> Exercise: Simulating evolution using the Price theorem</a>
<ul>
<li class="chapter" data-level="4.1" data-path="exercise-simulating-evolution-using-the-price-theorem.html"><a href="exercise-simulating-evolution-using-the-price-theorem.html#simulating-evolutionary-changes-in-allele-frequencies-1"><i class="fa fa-check"></i><b>4.1</b> Simulating evolutionary changes in allele frequencies</a></li>
<li class="chapter" data-level="4.2" data-path="exercise-simulating-evolution-using-the-price-theorem.html"><a href="exercise-simulating-evolution-using-the-price-theorem.html#the-regression-of-fitness-on-phenotypes-and-the-modes-of-selection"><i class="fa fa-check"></i><b>4.2</b> The regression of fitness on phenotypes and the modes of selection</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="exercise-simulating-evolution-using-landes-equation.html"><a href="exercise-simulating-evolution-using-landes-equation.html"><i class="fa fa-check"></i><b>5</b> Exercise: Simulating evolution using Lande’s equation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="exercise-simulating-evolution-using-landes-equation.html"><a href="exercise-simulating-evolution-using-landes-equation.html#simulating-evolutionary-changes-in-continuous-traits"><i class="fa fa-check"></i><b>5.1</b> Simulating evolutionary changes in continuous traits</a></li>
<li class="chapter" data-level="5.2" data-path="exercise-simulating-evolution-using-landes-equation.html"><a href="exercise-simulating-evolution-using-landes-equation.html#simulating-the-adaptive-landscape-of-a-continuous-trait"><i class="fa fa-check"></i><b>5.2</b> Simulating the adaptive landscape of a continuous trait</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to evolutionary and coevolutionary theory</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-the-course" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">2</span> Introduction to the course<a href="introduction-to-the-course.html#introduction-to-the-course" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<blockquote>
<p>“There is grandeur in this view of life, with its several powers, having been originally breathed into a few forms or into one; and that, whilst this planet has gone cycling on according to the fixed law of gravity, from so simple a beginning endless forms most beautiful and most wonderful have been, and are being, evolved.”</p>
</blockquote>
<p>The above quotation is one of the most iconic ones in biology. It beautifully concludes <em>The</em> <em>Origin of Species</em> by <em>Charles R. Darwin</em>, emphasizing the diversity of forms that we can find in nature, and how natural selection can shape these forms. Indeed, even in a simple walk through our neighborhood we can find hundreds of organisms, from plants to animals, with varying forms. From this observation, a natural question is then, how all these forms came to be?</p>
<p>To answer this question, we need to address two main problems. First, we need to assess the set of processes that act as a source of heritable variation on traits. Then, we need to understand how a second set of processes act upon the available heritable variation and shape traits. Our course focuses on this second problem. Specifically, we will study the theory and the mathematical foundations behind one of the main processes that can act upon heritable variation in populations: <strong>natural selection</strong>.</p>
<p>Throughout the first block, we will derive from first principles and gain some intuition about the main equations that predict how natural selection should drive evolution. In the second block, we will then use these equations to study ecological interactions as a source of reciprocal selective pressures that drive the coevolution between species.</p>
<div id="the-fundamental-concepts-of-evolution-by-natural-selection" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> The fundamental concepts of evolution by natural selection<a href="introduction-to-the-course.html#the-fundamental-concepts-of-evolution-by-natural-selection" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>First things first, we need a definition of evolution so that we can start from there. Following Futuyma (2005), evolution can be defined as a “change in the properties of groups of organisms over the course of generations”. Note that the <em>properties</em> in this definition of evolution “…embraces everything from slight changes in the proportions of different forms of a gene within a population to the alterations that led from the earliest organism to dinosaurs, bees, oaks, and humans” (Futuyma 2005). Throughout the course we will focus on how natural selection can drive the evolution of properties that can be measured as discrete or continuous traits in a population. Two biological examples of discrete and continuous traits would be the frequency of alleles that determine discrete floral morphs, or the length of floral tubes in population of plants, respectively. We will also focus on how these traits change within populations, or microevolutionary changes, in contrast with changes that could occur above the level of species, or macroevolutionary changes (Turner <em>et al.</em> 2009).</p>
<p>With this definition in hand, our first question is then <strong>how natural selection can drive changes in discrete or continuous traits in populations from one generation to the other?</strong> Before attempting to answer the question, we also need to define natural selection. Definitions are important because they allow us to build our models from fundamental propositions and assumptions that we know are true, or, the so called <em>first principles</em>. We can define natural selection as the differential survival and reproduction of individuals due to differences in heritable traits (Rice 2004). From this definition we can identify the three fundamental propositions that allow natural selection to operate:</p>
<ol style="list-style-type: decimal">
<li><p>Traits need to vary between individuals in a population.</p></li>
<li><p>These traits need to be heritable.</p></li>
<li><p>Individuals with different traits need to differ in how much they contribute with reproduction to the next generation.</p></li>
</ol>
<p>In evolutionary biology the word <em>fitness</em> is used to convey the reproductive contribution of an individual to the next generation (Rice 2004), and this is the definition of <em>fitness</em> that we will use throughout this course. Now that we have defined the scope of our problem (microevolution of discrete and continuous traits) and the first principles underlying our process of interest (the three propositions of natural selection), we can start to mathematically describe the problem.</p>
</div>
<div id="natural-selection-and-the-evolution-of-allele-frequencies-within-populations" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Natural selection and the evolution of allele frequencies within populations<a href="introduction-to-the-course.html#natural-selection-and-the-evolution-of-allele-frequencies-within-populations" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="building-our-first-equation" class="section level3 hasAnchor" number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> Building our first equation<a href="introduction-to-the-course.html#building-our-first-equation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Natural populations are complex, and it would be impossible to incorporate all the details that we can find when we mathematically describe a problem. Not only it would be impossible, but it would severely hinder our ability to understand and disentangle how our process of interest operates. Thus, our first step is to make some simplifying assumptions about these populations. Understanding what simplifying assumptions are reasonable is a complex topic and to learn more about this problem I highly recommend starting with the classic paper by Richard Levins, <em>The strategy of model building in population biology</em>. The short answer is that the choice of simplifying assumptions depends on your problem of interest and the questions that you are trying to answer.</p>
<p>Here, we are concerned about how populations evolve by means of natural selection. So, our first simplifying assumption is that the only mechanism of evolutionary change within a population is natural selection and that generations are non-overlapping. Our next step is to incorporate the fundamental propositions through which natural selection operates. The first one is that individuals vary in their traits. For these traits, we will assume the simplest case, in which a population contains individuals with only two discrete morphs/traits, for instance, plant individuals that produce only white flowers or purple flowers. From the second proposition, these traits need to be heritable. So, we will also assume that individuals are haploid, that two alleles, <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>, determine the two morphs (e.g. the white and purple morphs), and that these alleles are inherited by the offspring of individuals. Finally, each of the <span class="math inline">\(N_A\)</span> or <span class="math inline">\(N_B\)</span> individuals carrying the <span class="math inline">\(A\)</span> or <span class="math inline">\(B\)</span> alleles contributes to the next generation with an amount of <span class="math inline">\(W_A\)</span> and <span class="math inline">\(W_B\)</span> of individuals that inherit the <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> alleles, respectively. Thus, <span class="math inline">\(W_A\)</span> and <span class="math inline">\(W_B\)</span> represent the <em>fitness</em> of these individuals. Figure 1 below illustrates this scenario using the example of plant individuals displaying white or purple flowers.</p>
<div class="float">
<img src="figs/fig1.png" style="width:65.0%" alt="Fig. 1 - Theoretical example showing how the number of individuals of different morphs in a population can change from one generation to the other. For simplicity, in this example we assume that individuals are haploid and that two alleles, A and B, determine the white and purple floral morphs, respectively." />
<div class="figcaption">Fig. 1 - Theoretical example showing how the number of individuals of different morphs in a population can change from one generation to the other. For simplicity, in this example we assume that individuals are haploid and that two alleles, A and B, determine the white and purple floral morphs, respectively.</div>
</div>
<p><span class="math inline">\(\\\)</span></p>
<p>Under the above assumptions, in the next generation the total number of individuals carrying the allele <span class="math inline">\(A\)</span> in the population will depend on how much each individual contribute with offspring to the next generation. Mathematically, this is equivalent to multiply the current number of individuals carrying the allele <span class="math inline">\(A\)</span>, with their <em>fitness</em>, as follows:</p>
<p><span class="math display">\[N_{A}^{(t+1)}=N_{A}^{(t)}W_{A}^{(t)}\]</span></p>
<p>where the superscripts <span class="math inline">\((t)\)</span> and <span class="math inline">\((t+1)\)</span> denotes the current and the next generation, respectively. Similarly, the following equation describes the number of individuals carrying the <span class="math inline">\(B\)</span> allele in the next generation: <span class="math display">\[N_{B}^{(t+1)}=N_{B}^{(t)}W_{B}^{(t)}\]</span> However, we are interested in how the frequencies of these alleles change, not the absolute number of individuals. Since there are only two alleles in the population, if we find the frequency - <span class="math inline">\(p\)</span> - of allele <span class="math inline">\(A\)</span>, then the frequency of <span class="math inline">\(B\)</span> is equal to <span class="math inline">\((1-p)\)</span>. The frequency of <span class="math inline">\(A\)</span> in the next generation is equal to the number of individuals carrying <span class="math inline">\(A\)</span>, divided by total number of individuals in the next generation, <span class="math inline">\(\frac{N_{A}^{(t+1)}}{N^{(t+1)}}\)</span>. We already know what <span class="math inline">\(N_{A}^{(t+1)}\)</span> is, so our next step is to find what will be the total number of individuals in the next generation, <span class="math inline">\(N^{(t+1)}\)</span>. From the two equations above, we know that the total population size at the next generation is equal to:</p>
<p><span class="math display">\[N^{(t+1)}=N_{A}^{(t)}W_{A}^{(t)}+N_{B}^{(t)}W_{B}^{(t)}\]</span></p>
<p>We also know that at the current generation, the number of individuals carrying the allele <span class="math inline">\(A\)</span>, <span class="math inline">\(N_{A}^{(t)}\)</span>, is equal to the frequency of <span class="math inline">\(A\)</span>, multiplied by the total number of individuals in the population, that is <span class="math inline">\(N_{A}^{(t)}=p^{(t)}N^{(t)}\)</span>. In a similar way, <span class="math inline">\(N_{B}^{(t)}=(1-p^{(t)})N^{(t)}\)</span>. If we plug in these equivalences in the equation above, we get to the following one:</p>
<p><span class="math display">\[
\begin{aligned}
N^{(t+1)}&amp;=N_{A}^{(t)}W_{A}^{(t)}+N_{B}^{(t)}W_{B}^{(t)} \\
N^{(t+1)}&amp;=p^{(t)}N^{(t)}W_{A}^{(t)}+(1-p^{(t)})N^{(t)}W_{B}^{(t)}
\end{aligned}
\]</span></p>
<p>Further putting the <span class="math inline">\(N^{(t)}\)</span> that appears at the right-hand side in evidence leads to:</p>
<p><span class="math display">\[
\begin{aligned}
N^{(t+1)}&amp;=p^{(t)}N^{(t)}W_{A}^{(t)}+(1-p^{(t)})N^{(t)}W_{B}^{(t)} \\
N^{(t+1)}&amp;=N^{(t)}\left[p^{(t)}W_{A}^{(t)}+(1-p^{(t)})W_{B}^{(t)} \right] \\
N^{(t+1)}&amp;=N^{(t)}\overline{W}^{(t)}
\end{aligned}
\]</span> where <span class="math inline">\(\overline{W}^{(t)}=p^{(t)}W_{A}^{(t)}+(1-p^{(t)})W_{B}^{(t)}\)</span> is the average fitness of the population.</p>
<p><span class="math display">\[\begin{aligned}
\frac{N_{A}^{(t+1)}}{N^{(t+1)}}&amp;=\frac{N_{A}^{(t)}W_{A}^{(t)}}{N^{(t)}\overline{W}^{(t)}} \\
p^{(t+1)}&amp;=\frac{p^{(t)}W_{A}^{(t)}}{\overline{W}^{(t)}}
\end{aligned}\]</span></p>
<p>Now, lets try to understand what the equation above is telling us: at the next generation, the frequency of the allele <span class="math inline">\(A\)</span> in the population will depend on the current frequency of <span class="math inline">\(A\)</span>, multiplied by the fraction <span class="math inline">\(\frac{W_{A}^{(t)}}{\overline{W}^{(t)}}\)</span>. This fraction represents the fitness of <span class="math inline">\(A\)</span> relative to the average fitness in the population. On the one hand, if the fitness of <span class="math inline">\(A\)</span> is larger than the average of the population, the fraction is higher than 1, and the frequency of <span class="math inline">\(A\)</span> will increase in the next generation. On the other hand, when <span class="math inline">\(W_{A}^{(t)} &lt; \overline{W}^{(t)}\)</span>, the fraction will be smaller than 1 and the frequency of <span class="math inline">\(A\)</span> will decrease. What the equation is describing makes complete sense. If individuals of <span class="math inline">\(A\)</span> reproduce more than the average, then, the frequency of <span class="math inline">\(A\)</span> increases. This is exactly the way that we expect natural selection to operate, but now, we are able to <em>quantify</em> this expected outcome.</p>
</div>
<div id="wrights-equation-for-the-adaptive-landscape-of-alleles" class="section level3 hasAnchor" number="2.2.2">
<h3><span class="header-section-number">2.2.2</span> Wright’s equation for the adaptive landscape of alleles<a href="introduction-to-the-course.html#wrights-equation-for-the-adaptive-landscape-of-alleles" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can gain further insight about how selection drive the frequencies of alleles if we come up with an expression for the <em>change</em> in the frequency over generations, instead of just the frequency at the next generation. Mathematically, this is described by the difference <span class="math inline">\(p^{(t+1)} - p^{(t)}\)</span>, or <span class="math inline">\(\Delta p\)</span> . Plugging in the expression for <span class="math inline">\(p^{(t+1)}\)</span> that we just derived leads to the following equation:</p>
<p><span class="math display">\[\begin{aligned}
\Delta p&amp;=p^{(t+1)} - p^{(t)} \\
\Delta p&amp;=\frac{p^{(t)}W_{A}^{(t)}}{\overline{W}^{(t)}} - p^{(t)} \\
\Delta p&amp;=\frac{p^{(t)}W_{A}^{(t)}}{\overline{W}^{(t)}} - \frac{p^{(t)}\overline{W}^{(t)}}{\overline{W}^{(t)}} \\
\Delta p&amp;=p^{(t)}\frac{\left(W_{A}^{(t)}-\overline{W}^{(t)}\right)}{\overline{W}^{(t)}}
\end{aligned}\]</span></p>
<p>Substituting <span class="math inline">\(\overline{W}^{(t)}\)</span> in the numerator with <span class="math inline">\(p^{(t)}W_{A}^{(t)}+(1-p^{(t)})W_{B}^{(t)}\)</span> leads to:</p>
<p><span class="math display">\[\begin{aligned}
\Delta p&amp;=p^{(t)}\frac{\left[W_{A}^{(t)}-p^{(t)}W_{A}^{(t)}-(1-p^{(t)})W_{B}^{(t)}\right]}{\overline{W}^{(t)}} \\
\Delta p&amp;=p^{(t)}\frac{\left[(1-p^{(t)})W_{A}^{(t)}-(1-p^{(t)})W_{B}^{(t)}\right]}{\overline{W}^{(t)}} \\
\Delta p&amp;=p^{(t)}(1-p^{(t)})\frac{\left(W_{A}^{(t)}-W_{B}^{(t)}\right)}{\overline{W}^{(t)}}
\end{aligned}\]</span></p>
<p>This equation show that the change in the frequency of the <span class="math inline">\(A\)</span> allele depends on two components. First, it depends on <span class="math inline">\(p^{(t)}(1-p^{(t)})\)</span>. This term is the product of the frequencies of <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>, which, from probability theory, represents the variance in frequencies. From this interpretation, the term <span class="math inline">\(p^{(t)}(1-p^{(t)})\)</span> is the amount of genetic variance within the population. Second, the change in frequency also depends on the term <span class="math inline">\(\frac{\left(W_{A}-W_{B}^{(t)}\right)}{\overline{W}^{(t)}}\)</span>, representing the difference in fitness between individuals carrying the <span class="math inline">\(A\)</span> or <span class="math inline">\(B\)</span> alleles. When <span class="math inline">\(W_{A}&gt;W_{B}\)</span>, this term is positive and it is negative when <span class="math inline">\(W_{A}&lt;W_{B}\)</span>. Therefore, the equation shows that selection drives evolutionary changes depending on the amount of the available genetic variance within the population and differences in fitness. On the one hand genetic variance scales the amount of evolutionary change that can occur, such that if there is no genetic variance, i.e., <span class="math inline">\(p\neq 1\)</span> and <span class="math inline">\(p\neq 0\)</span>, no change occurs. On the other hand, differences in fitness control the direction of the evolutionary changes. Since the variance is always a positive number, when <span class="math inline">\(W_{A}&gt;W_{B}\)</span>, <span class="math inline">\(\Delta p\)</span> is also positive and the frequency of <span class="math inline">\(p\)</span> increases. In contrast, when <span class="math inline">\(W_{A}&lt;W_{B}\)</span>, <span class="math inline">\(\Delta p\)</span> will be negative and the frequency of <span class="math inline">\(p\)</span> decreases. If there are no differences in fitness, or <span class="math inline">\(W_{A} = W_{B}\)</span>, there is no evolutionary change. Once more, this behavior makes complete sense from what we would expect from evolution by natural selection.</p>
<p>The equation above for the change in allele frequencies was the one derived by Sewall Wright in 1937 in the classic paper <em>The distribution of gene frequencies in populations</em>. The key idea of the paper was identifying that the term <span class="math inline">\(\left(W_{A}^{(t)}-W_{B}^{(t)}\right)\)</span> corresponds to the derivative of the average fitness of the population, relative to the frequency of allele <span class="math inline">\(p\)</span>, as follows:</p>
<p><span class="math display">\[\begin{aligned}
\frac{\partial \overline{W}^{(t)}}{\partial p} &amp;= \frac{\partial}{\partial p^{(t)}}p^{(t)}W_{A}^{(t)} + \frac{\partial}{\partial p^{(t)}}(1-p^{(t)})W_{B}^{(t)} \\
\frac{\partial \overline{W}^{(t)}}{\partial p^{(t)}} &amp;=  W_{A}^{(t)} - W_{B}^{(t)}
\end{aligned}\]</span></p>
<p>This result holds whenever the <em>fitness</em> of alleles does not depend on their frequencies, or, in other words, when <em>fitness</em> is frequency independent. The result above allow us to rewrite the equation for the change in allele frequencies as follows:</p>
<p><span class="math display">\[
\Delta p=p^{(t)}(1-p^{(t)})\frac{\partial ln\overline{W}^{(t)}}{\partial p^{(t)}}
\]</span> where the <em>ln</em> term in <span class="math inline">\(\frac{\partial ln\overline{W}^{(t)}}{\partial p^{(t)}}\)</span> comes from the calculus rule that <span class="math inline">\(\frac{1}{x}\frac{d}{dx} = \frac{dlnx}{dx}\)</span>.</p>
<p>Rewriting the equation in terms of this derivative is key to understand the concept of how natural selection drive adaptations in populations. A derivative represents the slope of a curve at a point. In turn, the average <em>fitness</em> of a population represents its growth rate. Thus, we can think about the derivative of the average fitness relative to the frequency of an allele as a population “climbing” a slope that is defined by the growth rate of the population. Visually this can be represented with the relationship between the average fitness of the population and the frequency of a given allele, as the figure below shows:</p>
<div class="float">
<img src="figs/fig2.png" style="width:50.0%" alt="Fig. 2 - Hypothetical example of the relationship between the average fitness of a population and the frequency of an allele (p). The dashed line represents the point at which the slope (the derivative) is equal to 0. In the points to the left of the dashed line, the slope is positive, and selection favors an increase in the frequency of the allele. In contrast, the slope is negative in the points to the right of the dashed line and selection favors a decrease in the frequency of the allele. Importantly, across all points selection will drive the frequencies of the allele in the direction at which the average fitness of the population increases." />
<div class="figcaption">Fig. 2 - Hypothetical example of the relationship between the average fitness of a population and the frequency of an allele (p). The dashed line represents the point at which the slope (the derivative) is equal to 0. In the points to the left of the dashed line, the slope is positive, and selection favors an increase in the frequency of the allele. In contrast, the slope is negative in the points to the right of the dashed line and selection favors a decrease in the frequency of the allele. Importantly, across all points selection will drive the frequencies of the allele in the direction at which the average fitness of the population increases.</div>
</div>
<p><span class="math inline">\(\\\)</span></p>
<p>The figure above illustrates the concept introduced by Wright (1937) of an <em>adaptive landscape</em>. In evolutionary biology, the <em>adaptive landscape</em> depicts the relationship between the average fitness of a population and the values of traits, which, in our example, are the frequencies of alleles. From the figure and the interpretation of the derivative outlined above, we can see the natural selection will drive the evolution of allele frequencies in the direction that increases the average fitness of the population. This provides a mathematical formulation and proof of the intuitive concept of how natural selection favors <em>adaptations</em> in populations. This relationship always holds whenever <em>fitness</em> is frequency-independent, but not when <em>fitness</em> is frequency-dependent as we will see next in the exercise session.</p>
<p>Going back to our initial question: <strong>how natural selection can drive changes in discrete or continuous traits in populations from one generation to the other?</strong> Using Wright’s equation we now can partially answer this question. For discrete traits, for instance, alleles, we learned that selection will favor the evolution of an increased frequency of alleles that maximize the average <em>fitness</em> of the population, leading to <em>adaptations</em>. The rate at which the population evolves will depend on the available genetic variation in the population, as well as how large are the differences in <em>fitness</em> between individuals carrying different alleles.</p>
</div>
<div id="references" class="section level3 hasAnchor" number="2.2.3">
<h3><span class="header-section-number">2.2.3</span> References<a href="introduction-to-the-course.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Darwin, Charles Robert. 1861. <em>On the Origin of Species by Means of Natural Selection, or the Preservation of Favoured Races in the Struggle for Life</em>. 3d edition. London: John Murray.</p>
<p>Futuyma, Douglas. 2005. <em>Evolution</em>. Sunderland, MA: Sinauer Associates.</p>
<p>Levins, R. 1966. The strategy of model building in population biology. American Scientist 54 (4): 421–31.</p>
<p>Rice, Sean H. 2004. <em>Evolutionary Theory: Mathematical and Conceptual Foundations</em>. New York, NY: Oxford University Press.</p>
<p>Turner, Derek and Joyce C. Havstad. 2019. Philosophy of Macroevolution. <em>The Stanford Encyclopedia of Philosophy</em> (Summer 2019 Edition), Edward N. Zalta (ed.). Available in: <a href="https://plato.stanford.edu/archives/sum2019/entries/macroevolution" class="uri">https://plato.stanford.edu/archives/sum2019/entries/macroevolution</a>.</p>
<p>Wright, S. 1937. The distribution of gene frequencies in populations. <em>Proceedings of the National Academy of Sciences of the United States of America</em> 23 (6): 307–20.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="exercise-evolution-of-allele-frequencies.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/01-intro.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
